skbio/sequence/_sequence.py

```diff
@@ -187,30 +351,107 @@ def metadata(self):
-        ...              positional_metadata={'quality': range(16)})
+        >>> s = Sequence(
+        ...     'ACGT',
+        ...     positional_metadata={'quality': [3, 3, 20, 11],
+        ...                          'exons': [True, True, False, True]})
         >>> s # doctest: +NORMALIZE_WHITESPACE
-        Sequence('ACGTACGTACGTACGT', length=16, has_metadata=False,
+        Sequence('ACGT', length=4, has_metadata=False,
                  has_positional_metadata=True)
-        >>> s.positional_metadata['quality'].values
-        array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15])
+
+        Retrieve positional metadata:
+
+        >>> s.positional_metadata
+           exons  quality
+        0   True        3
+        1   True        3
+        2  False       20
+        3   True       11
+
+        Update positional metadata:
+
+        >>> s.positional_metadata['gaps'] = [False, False, False, False]
+        >>> s.positional_metadata
+           exons  quality   gaps
+        0   True        3  False
+        1   True        3  False
+        2  False       20  False
+        3   True       11  False
+
+        Set positional metadata:
+
+        >>> s.positional_metadata = {'degenerates':
+        ...                          [False, False, False, False]}
+        >>> s.positional_metadata
+          degenerates
+        0       False
+        1       False
+        2       False
+        3       False
+
+        Delete positional metadata:
+
+        >>> s.has_positional_metadata()
+        True
+        >>> del s.positional_metadata
+        >>> s.positional_metadata
+        Empty DataFrame
+        Columns: []
+        Index: [0, 1, 2, 3]
+        >>> s.has_positional_metadata()
+        False
 
         """
         return self._positional_metadata
 
+    @positional_metadata.setter
+    def positional_metadata(self, positional_metadata):
+        try:
+            # copy=True to copy underlying data buffer
+            positional_metadata = pd.DataFrame(positional_metadata, copy=True)
+        except pd.core.common.PandasError as e:
+            raise TypeError('Positional metadata invalid. Must be consumable '
+                            'by pandas.DataFrame. Original pandas error '
+                            'message: "%s"' % e)
+
+        num_rows = len(positional_metadata.index)
+        if num_rows != len(self):
+            raise ValueError(
+                "Number of positional metadata values (%d) must match the "
+                "number of characters in the sequence (%d)." %
+                (num_rows, len(self)))
+
+        positional_metadata.reset_index(drop=True, inplace=True)
+        self._positional_metadata = positional_metadata
+
+    @positional_metadata.deleter
+    def positional_metadata(self):
+        if hasattr(self, '_positional_metadata'):
+            del self._positional_metadata
+        self.positional_metadata = pd.DataFrame(index=np.arange(len(self)))
```

Comment:
Any chance this can be optimized to be `self._positional_metadata = None` and the getter can just build a dataframe on the fly? You shouldn't have to change any tests